### PR TITLE
Add CodeObject::parse overloads to accept multiple targets

### PR DESCRIPTION
**Pull request by batuzovk**
Parser::parse_at takes only one target at a time. This has two repercussions:

* It can't take full advantage of parallel parsing when there are multiple targets in need of it.
* It still enters and exits OMP parallel loop. This operation has its own performance overhead that depends on the host system (number of processors and OMP configuration). This overhead can be very significant.

This change introduces Parser::parse_vec that takes a vector of targets and parses them in parallel.

When there are significant number of parse_at calls, performance gains from migrating to parse_vec can be huge. Parsing what initial parse() missed in libLLVM-12.so.1 with parse_at took nearly 20 minutes on 4-core hyper-threaded CPU. Doing the same with parse_vec took less than 4 minutes. On 128-bit ARM64 server we've seen even bigger gains (from around 40 minutes to less than 3).

### CODE REVIEW

**Comment by batuzovk** [comment]
Done. By the way, when I update branch via github web interface it runs CI automatically without waiting for approval from authorized users. Is this intended?

**Comment by kupsch** [comment]
It was unintended (set to only require approval for first time committers).  Fixed now.

**Review by hainest — CHANGES_REQUESTED** [review]
Overall, this is a good feature to have with a few minor edits. However, we're concerned about the parsing issues that led to its necessity. On my local machine, I can parse libLLVM.so.15 in about a second. It finds ~35000 functions and readelf reports ~30000 non-weak function symbols. I'd say we got pretty much everything. I have a few questions about your experience.

1. How many OMP threads are you using?
2. Is your libLLVM stripped?
3. What targets are you needing to explicitly parse with `parse_at`? How did you create that list (from DWARF, some known list, etc.)?
4. Taking upwards of 40 mins to parse libLLVM (~112MB for me) is a serious problem. The serial overhead of calling `parse_at` multiple times shouldn't be a major bottleneck unless it gets called thousands of times. How many extra targets were you needing to parse?

**Comment by batuzovk** [comment]
> Overall, this is a good feature to have with a few minor edits. However, we're concerned about the parsing issues that led to its necessity. On my local machine, I can parse libLLVM.so.15 in about a second. It finds ~35000 functions and readelf reports ~30000 non-weak function symbols. I'd say we got pretty much everything. I have a few questions about your experience.
> 
>     1. How many OMP threads are you using?
> 
>     2. Is your libLLVM stripped?
> 
>     3. What targets are you needing to explicitly parse with `parse_at`? How did you create that list (from DWARF, some known list, etc.)?
> 
>     4. Taking upwards of 40 mins to parse libLLVM (~112MB for me) is a serious problem. The serial overhead of calling `parse_at` multiple times shouldn't be a major bottleneck unless it gets called thousands of times. How many extra targets were you needing to parse?

1. We didn't limit number of OMP threads. The server we've run it on has 128 cores, so I guess the number of threads was the same.
2. Yes, it is stripped.
3. A whole bunch of targets, really. Some are discovered by examining relocations, some - with dataflow analysis.
4. It was called tens of thousands of times actually. I've rerun our tool on libLLVM-12.so.1. Initial parse() finds ~43600 functions, after our analysis we get ~78600 functions. So, 35000 more functions. Since some are found recursively, I'd assume we had something like 20000 to 25000 parse_at() calls. OMP functions occupied more than 95% of the pref report.

> It finds ~35000 functions and readelf reports ~30000 non-weak function symbols. I'd say we got pretty much everything.

It isn't about function that have symbols. Those will be in hints and are guaranteed to be discovered. It is about functions that don't have symbols associated with them. There are plenty of those. For example, on stripped libLLVM-12.so.1 from our example llvm-bolt finds over 80000 functions. So, even our analysis is few thousands functions short, and initial parse() covered about half of it. Symbol table for that library contains 41284 entries, for reference.

If you want to know how well dyninst has covered this or that binary you usually can look at gaps in .text. If you see something bigger than 32 bytes on x86_64 binary, then chances are there is a function dyninst has missed there. Data inside code segment happens, but it is not common in general libraries you'll find under /usr/lib of your Linux distribution. And if the gap is several kilobytes in size, then there are probably more than one function hidden there.

I'll update code according to your suggestions tomorrow.

File: parseAPI/src/Parser.C
Comment (batuzovk) [reply parseAPI/src/Parser.C]: I ended up with 4 variants: 

1. Address - forwards to 3,
2. pair<Address,CodeRegion*> - forwards to 4,
3. vector<Address> - finds regions and forwards to 4,
4. vector<pair<Address,CodeRegion*>> - does actual parsing.

No code duplication any more here, at least not in any substantial way.

**Comment by hainest** [comment]
LGTM. @kupsch ?

**Review by kupsch — APPROVED** [review]
Also LGTM.

**Review by hainest — APPROVED** [review]

**Pull request merged**
